File: repository/rule.go
package repository

import (
	"context"
	"database/sql"
	"fmt"
	"strings"
	"time"

	"github.com/mohan9182/financer/api"
	"github.com/google/uuid"
)

type RuleRepository struct {
	*BaseRepository
}

func NewRuleRepository(base *BaseRepository) *RuleRepository {
	return &RuleRepository{BaseRepository: base}
}

type RuleRecord struct {
	ID        string
	Name      string
	Priority  int
	Logic     string
	CreatedAt time.Time
}

type RuleCondData struct {
	MatchField string
	Operator   string
	Pattern    string
}

func scanRule(row scannable) (*api.RuleResponse, error) {
	var rule api.RuleResponse
	var createdAt time.Time
	var logicStr string
	err := row.Scan(&rule.Id, &rule.Name, &rule.Priority, &logicStr, &createdAt)
	if err != nil {
		return nil, err
	}
	rule.Logic = parseLogicEnum(logicStr)
	rule.CreatedAt = createdAt
	return &rule, nil
}

func (r *RuleRepository) Create(ctx context.Context, userID, name string, priority int32, logic string, conditions []RuleCondData, actions []*api.RuleAction) (*api.RuleResponse, error) {
	id := uuid.New().String()
	now := time.Now().UTC()

	err := r.ExecInTx(ctx, func(tx *Tx) error {
		if _, err := tx.ExecContext(ctx,
			`INSERT INTO rules (id, name, priority, logic, created_at, user_id) VALUES (?, ?, ?, ?, ?, ?)`,
			id, name, priority, logic, now, userID,
		); err != nil {
			return fmt.Errorf("inserting rule: %w", err)
		}

		for _, cond := range conditions {
			condID := uuid.New().String()
			if _, err := tx.ExecContext(ctx,
				`INSERT INTO rule_conditions (id, rule_id, match_field, operator, pattern) VALUES (?, ?, ?, ?, ?)`,
				condID, id, cond.MatchField, cond.Operator, cond.Pattern,
			); err != nil {
				return fmt.Errorf("inserting rule condition: %w", err)
			}
		}

		if err := r.insertActions(ctx, tx, id, actions); err != nil {
			return err
		}
		return nil
	})
	if err != nil {
		return nil, err
	}

	return &api.RuleResponse{
		Id:         id,
		Name:       name,
		Priority:   priority,
		Logic:      parseLogicEnum(logic),
		Conditions: buildConditionProtos(conditions),
		Actions:    actions,
		CreatedAt:  now,
	}, nil
}

func (r *RuleRepository) insertActions(ctx context.Context, tx *Tx, id string, actions []*api.RuleAction) error {
	for _, act := range actions {
		actionID := uuid.New().String()
		actionType := "SET_NAME"
		var nameOp, value, catID, transferAccountID string
		if act.SetTransferAccountId != "" {
			actionType = "SET_TRANSFER_ACCOUNT"
			transferAccountID = act.SetTransferAccountId
		} else if act.SetCategoryId != "" {
			actionType = "SET_CATEGORY"
			catID = act.SetCategoryId
		} else {
			nameOp = act.SetNameOp.String()
			value = act.SetName
		}
		if _, err := tx.ExecContext(ctx,
			`INSERT INTO rule_actions (id, rule_id, action_type, name_op, value, category_id, transfer_account_id) VALUES (?, ?, ?, ?, ?, ?, ?)`,
			actionID, id, actionType, nameOp, value, catID, transferAccountID,
		); err != nil {
			return fmt.Errorf("inserting rule action: %w", err)
		}
	}
	return nil
}

func (r *RuleRepository) GetByID(ctx context.Context, id string) (*api.RuleResponse, error) {
	rule, err := QueryOne(ctx, r.readDB,
		`SELECT id, name, priority, logic, created_at FROM rules WHERE id = ?`,
		scanRule, id,
	)
	if err != nil {
		return nil, fmt.Errorf("querying rule: %w", err)
	}
	if rule == nil {
		return nil, nil
	}

	rule.Conditions, _ = r.GetConditions(ctx, id)
	rule.Actions, _ = r.GetActions(ctx, id)
	return rule, nil
}

func (r *RuleRepository) Update(ctx context.Context, id, name string, priority int32, logic string, conditions []RuleCondData, actions []*api.RuleAction) error {
	return r.ExecInTx(ctx, func(tx *Tx) error {
		result, err := tx.ExecContext(ctx,
			`UPDATE rules SET name = COALESCE(NULLIF(?, ''), name), priority = ?, logic = ? WHERE id = ?`,
			name, priority, logic, id,
		)
		if err != nil {
			return fmt.Errorf("updating rule: %w", err)
		}

		rows, _ := result.RowsAffected()
		if rows == 0 {
			return sql.ErrNoRows
		}

		if _, err := tx.ExecContext(ctx, `DELETE FROM rule_conditions WHERE rule_id = ?`, id); err != nil {
			return fmt.Errorf("deleting old conditions: %w", err)
		}

		for _, cond := range conditions {
			condID := uuid.New().String()
			if _, err := tx.ExecContext(ctx,
				`INSERT INTO rule_conditions (id, rule_id, match_field, operator, pattern) VALUES (?, ?, ?, ?, ?)`,
				condID, id, cond.MatchField, cond.Operator, cond.Pattern,
			); err != nil {
				return fmt.Errorf("inserting condition: %w", err)
			}
		}

		if _, err := tx.ExecContext(ctx, `DELETE FROM rule_actions WHERE rule_id = ?`, id); err != nil {
			return fmt.Errorf("deleting old actions: %w", err)
		}

		return r.insertActions(ctx, tx, id, actions)
	})
}

func (r *RuleRepository) Delete(ctx context.Context, id string) (bool, error) {
	result, err := r.ExecContext(ctx, `DELETE FROM rules WHERE id = ?`, id)
	if err != nil {
		return false, fmt.Errorf("deleting rule: %w", err)
	}

	rows, _ := result.RowsAffected()
	return rows > 0, nil
}

func (r *RuleRepository) List(ctx context.Context, userID string) ([]*api.RuleResponse, error) {
	results, err := QueryAll(ctx, r.readDB,
		`SELECT id, name, priority, logic, created_at FROM rules WHERE user_id = ? ORDER BY priority DESC, name ASC`,
		scanRule, userID,
	)
	if err != nil {
		return nil, fmt.Errorf("listing rules: %w", err)
	}

	ids := make([]string, len(results))
	for i, rule := range results {
		ids[i] = rule.Id
	}
	conds, _ := r.conditionsByRule(ctx, ids)
	acts, _ := r.actionsByRule(ctx, ids)
	for _, rule := range results {
		rule.Conditions = conds[rule.Id]
		rule.Actions = acts[rule.Id]
	}
	return results, nil
}

func inPlaceholders(n int) string {
	ps := make([]string, n)
	for i := range ps {
		ps[i] = "?"
	}
	return strings.Join(ps, ",")
}

func (r *RuleRepository) conditionsByRule(ctx context.Context, ids []string) (map[string][]*api.RuleCondition, error) {
	out := make(map[string][]*api.RuleCondition)
	if len(ids) == 0 {
		return out, nil
	}
	rows, err := r.readDB.QueryContext(ctx,
		`SELECT rule_id, match_field, operator, pattern FROM rule_conditions WHERE rule_id IN (`+inPlaceholders(len(ids))+`)`, toAny(ids)...)
	if err != nil {
		return nil, fmt.Errorf("querying conditions: %w", err)
	}
	defer rows.Close()
	for rows.Next() {
		var rid, mf, op, pattern string
		if err := rows.Scan(&rid, &mf, &op, &pattern); err != nil {
			return nil, fmt.Errorf("scanning condition: %w", err)
		}
		out[rid] = append(out[rid], &api.RuleCondition{
			MatchField: parseMatchFieldEnum(mf),
			Operator:   parseMatchOperatorEnum(op),
			Pattern:    pattern,
		})
	}
	return out, rows.Err()
}

func (r *RuleRepository) actionsByRule(ctx context.Context, ids []string) (map[string][]*api.RuleAction, error) {
	out := make(map[string][]*api.RuleAction)
	if len(ids) == 0 {
		return out, nil
	}
	rows, err := r.readDB.QueryContext(ctx,
		`SELECT rule_id, action_type, name_op, value, category_id, transfer_account_id FROM rule_actions
		 WHERE rule_id IN (`+inPlaceholders(len(ids))+`) ORDER BY rule_id, rowid`, toAny(ids)...)
	if err != nil {
		return nil, fmt.Errorf("querying actions: %w", err)
	}
	defer rows.Close()
	for rows.Next() {
		var rid, actionType, nameOp, value, catID, transferAccountID string
		if err := rows.Scan(&rid, &actionType, &nameOp, &value, &catID, &transferAccountID); err != nil {
			return nil, fmt.Errorf("scanning action: %w", err)
		}
		a := &api.RuleAction{}
		if actionType == "SET_TRANSFER_ACCOUNT" {
			a.SetTransferAccountId = transferAccountID
		} else if actionType == "SET_CATEGORY" {
			a.SetCategoryId = catID
		} else {
			a.SetName = value
			a.SetNameOp = parseRuleActionOp(nameOp)
		}
		out[rid] = append(out[rid], a)
	}
	return out, rows.Err()
}

func toAny(ids []string) []any {
	out := make([]any, len(ids))
	for i, id := range ids {
		out[i] = id
	}
	return out
}

func (r *RuleRepository) GetConditions(ctx context.Context, ruleID string) ([]*api.RuleCondition, error) {
	rows, err := r.readDB.QueryContext(ctx,
		`SELECT match_field, operator, pattern FROM rule_conditions WHERE rule_id = ?`, ruleID)
	if err != nil {
		return nil, fmt.Errorf("querying conditions: %w", err)
	}
	defer rows.Close()

	var conditions []*api.RuleCondition
	for rows.Next() {
		var c api.RuleCondition
		var matchField, operator string
		if err := rows.Scan(&matchField, &operator, &c.Pattern); err != nil {
			return nil, fmt.Errorf("scanning condition: %w", err)
		}
		c.MatchField = parseMatchFieldEnum(matchField)
		c.Operator = parseMatchOperatorEnum(operator)
		conditions = append(conditions, &c)
	}
	if rows.Err() != nil {
		return nil, rows.Err()
	}
	return conditions, nil
}

func (r *RuleRepository) GetActions(ctx context.Context, ruleID string) ([]*api.RuleAction, error) {
	rows, err := r.readDB.QueryContext(ctx,
		`SELECT action_type, name_op, value, category_id, transfer_account_id FROM rule_actions WHERE rule_id = ? ORDER BY rowid`, ruleID)
	if err != nil {
		return nil, fmt.Errorf("querying actions: %w", err)
	}
	defer rows.Close()
	var actions []*api.RuleAction
	for rows.Next() {
		var actionType, nameOp, value, catID, transferAccountID string
		if err := rows.Scan(&actionType, &nameOp, &value, &catID, &transferAccountID); err != nil {
			return nil, fmt.Errorf("scanning action: %w", err)
		}
		a := &api.RuleAction{}
		if actionType == "SET_TRANSFER_ACCOUNT" {
			a.SetTransferAccountId = transferAccountID
		} else if actionType == "SET_CATEGORY" {
			a.SetCategoryId = catID
		} else {
			a.SetName = value
			a.SetNameOp = parseRuleActionOp(nameOp)
		}
		actions = append(actions, a)
	}
	return actions, rows.Err()
}

type OverlayRule struct {
	Logic      string
	Actions    []*api.RuleAction
	Conditions []RuleCondData
}

// ListForOverlay loads a user's rules ordered by priority (highest first) with
// their conditions and actions, for read-time compound mapping.
func (r *RuleRepository) ListForOverlay(ctx context.Context, userID string) ([]OverlayRule, error) {
	records, err := r.List(ctx, userID)
	if err != nil {
		return nil, err
	}
	// List orders by priority DESC already.
	rules := make([]OverlayRule, 0, len(records))
	for _, rec := range records {
		rules = append(rules, OverlayRule{
			Logic:      LogicString(rec.Logic),
			Actions:    rec.Actions,
			Conditions: ConditionsToData(rec.Conditions),
		})
	}
	return rules, nil
}

func ConditionsToData(conds []*api.RuleCondition) []RuleCondData {
	out := make([]RuleCondData, 0, len(conds))
	for _, c := range conds {
		out = append(out, RuleCondData{
			MatchField: c.MatchField.String(),
			Operator:   c.Operator.String(),
			Pattern:    c.Pattern,
		})
	}
	return out
}

func parseRuleActionOp(s string) api.RuleActionOp {
	switch s {
	case "RULE_ACTION_OP_RENAME":
		return api.RuleActionOp_RULE_ACTION_OP_RENAME
	case "RULE_ACTION_OP_ADD_PREFIX":
		return api.RuleActionOp_RULE_ACTION_OP_ADD_PREFIX
	case "RULE_ACTION_OP_ADD_SUFFIX":
		return api.RuleActionOp_RULE_ACTION_OP_ADD_SUFFIX
	default:
		return api.RuleActionOp_RULE_ACTION_OP_UNSPECIFIED
	}
}

func parseLogicEnum(s string) api.RuleLogic {
	switch s {
	case "AND":
		return api.RuleLogic_RULE_LOGIC_AND
	default:
		return api.RuleLogic_RULE_LOGIC_OR
	}
}

func parseMatchFieldEnum(s string) api.RuleMatchField {
	switch s {
	case "RULE_MATCH_FIELD_NAME":
		return api.RuleMatchField_RULE_MATCH_FIELD_NAME
	case "RULE_MATCH_FIELD_AMOUNT":
		return api.RuleMatchField_RULE_MATCH_FIELD_AMOUNT
	case "RULE_MATCH_FIELD_TYPE":
		return api.RuleMatchField_RULE_MATCH_FIELD_TYPE
	case "RULE_MATCH_FIELD_CATEGORY":
		return api.RuleMatchField_RULE_MATCH_FIELD_CATEGORY
	case "RULE_MATCH_FIELD_ACCOUNT":
		return api.RuleMatchField_RULE_MATCH_FIELD_ACCOUNT
	default:
		return api.RuleMatchField_RULE_MATCH_FIELD_UNSPECIFIED
	}
}

func parseMatchOperatorEnum(s string) api.RuleMatchOperator {
	switch s {
	case "RULE_MATCH_OPERATOR_CONTAINS":
		return api.RuleMatchOperator_RULE_MATCH_OPERATOR_CONTAINS
	case "RULE_MATCH_OPERATOR_STARTS_WITH":
		return api.RuleMatchOperator_RULE_MATCH_OPERATOR_STARTS_WITH
	case "RULE_MATCH_OPERATOR_ENDS_WITH":
		return api.RuleMatchOperator_RULE_MATCH_OPERATOR_ENDS_WITH
	case "RULE_MATCH_OPERATOR_EQUALS":
		return api.RuleMatchOperator_RULE_MATCH_OPERATOR_EQUALS
	case "RULE_MATCH_OPERATOR_GREATER_THAN":
		return api.RuleMatchOperator_RULE_MATCH_OPERATOR_GREATER_THAN
	case "RULE_MATCH_OPERATOR_LESS_THAN":
		return api.RuleMatchOperator_RULE_MATCH_OPERATOR_LESS_THAN
	case "RULE_MATCH_OPERATOR_REGEX":
		return api.RuleMatchOperator_RULE_MATCH_OPERATOR_REGEX
	default:
		return api.RuleMatchOperator_RULE_MATCH_OPERATOR_UNSPECIFIED
	}
}

func buildConditionProtos(conditions []RuleCondData) []*api.RuleCondition {
	var result []*api.RuleCondition
	for _, c := range conditions {
		result = append(result, &api.RuleCondition{
			MatchField: parseMatchFieldEnum(c.MatchField),
			Operator:   parseMatchOperatorEnum(c.Operator),
			Pattern:    c.Pattern,
		})
	}
	return result
}

func LogicString(logic api.RuleLogic) string {
	if logic == api.RuleLogic_RULE_LOGIC_AND {
		return "AND"
	}
	return "OR"
}

File: src/repo/sqlite/rule.rs
//! `SQLite` rules repository — SQL + read-time overlay data, implementing
//! `crate::repo::traits::RuleRepo` over an `sqlx::SqlitePool`.

use async_trait::async_trait;
use sqlx::SqlitePool;
use uuid::Uuid;

use crate::error::Result;
use crate::repo::traits::rule::{ActionOp, ActionType, ConditionData, MatchField, MatchOperator, OverlayRule, Rule, RuleAction, RuleCondition, RuleLogic};
use crate::utils::timex::{ts_rfc3339};

pub struct SqliteRuleRepo {
    pool: SqlitePool,
}

impl SqliteRuleRepo {
    pub fn new(pool: SqlitePool) -> Self {
        Self { pool }
    }

    async fn create_inner(
        &self,
        user_id: &str,
        name: &str,
        priority: i64,
        logic: RuleLogic,
        conditions: &[RuleCondition],
        actions: &[RuleAction],
    ) -> Result<Rule> {
        let id = Uuid::new_v4().to_string();
        let now = crate::utils::timex::now_go_ts();
        let mut tx = self.pool.begin().await?;
        sqlx::query("INSERT INTO rules (id, name, priority, logic, created_at, user_id) VALUES (?, ?, ?, ?, ?, ?)")
            .bind(&id)
            .bind(name)
            .bind(priority)
            .bind(logic.to_string())
            .bind(&now)
            .bind(user_id)
            .execute(&mut *tx)
            .await?;
        insert_conditions(&mut tx, &id, conditions).await?;
        insert_actions(&mut tx, &id, actions).await?;
        tx.commit().await?;
        Ok(Rule { id, name: name.to_string(), priority, logic, conditions: conditions.to_vec(), actions: actions.to_vec(), created_at: ts_rfc3339(&now) })
    }

    #[allow(clippy::too_many_arguments)]
    async fn update_inner(
        &self,
        user_id: &str,
        id: &str,
        name: &str,
        priority: i64,
        logic: RuleLogic,
        conditions: &[RuleCondition],
        actions: &[RuleAction],
    ) -> Result<bool> {
        let mut tx = self.pool.begin().await?;
        let result = sqlx::query("UPDATE rules SET name = COALESCE(NULLIF(?, ''), name), priority = ?, logic = ? WHERE id = ? AND user_id = ?")
            .bind(name)
            .bind(priority)
            .bind(logic.to_string())
            .bind(id)
            .bind(user_id)
            .execute(&mut *tx)
            .await?;
        if result.rows_affected() == 0 {
            return Ok(false);
        }
        sqlx::query("DELETE FROM rule_conditions WHERE rule_id = ?").bind(id).execute(&mut *tx).await?;
        sqlx::query("DELETE FROM rule_actions WHERE rule_id = ?").bind(id).execute(&mut *tx).await?;
        insert_conditions(&mut tx, id, conditions).await?;
        insert_actions(&mut tx, id, actions).await?;
        tx.commit().await?;
        Ok(true)
    }

    async fn delete_inner(&self, user_id: &str, id: &str) -> Result<bool> {
        let result = sqlx::query("DELETE FROM rules WHERE id = ? AND user_id = ?")
            .bind(id)
            .bind(user_id)
            .execute(&self.pool)
            .await?;
        Ok(result.rows_affected() > 0)
    }

    async fn get_by_id_inner(&self, user_id: &str, id: &str) -> Result<Option<Rule>> {
        let row = sqlx::query_as::<_, RawRule>("SELECT id, name, priority, logic, created_at FROM rules WHERE id = ? AND user_id = ?")
            .bind(id)
            .bind(user_id)
            .fetch_optional(&self.pool)
            .await?;
        let Some(raw) = row else { return Ok(None) };
        let conditions = self.conditions_for_rule(id).await?;
        let actions = self.actions_for_rule(id).await?;
        Ok(Some(raw.into_rule(conditions, actions)))
    }

    async fn list_inner(&self, user_id: &str) -> Result<Vec<Rule>> {
        let rows = sqlx::query_as::<_, RawRule>(
            "SELECT id, name, priority, logic, created_at FROM rules WHERE user_id = ? ORDER BY priority DESC, name ASC",
        )
        .bind(user_id)
        .fetch_all(&self.pool)
        .await?;
        let mut rules = Vec::new();
        for raw in rows {
            let conditions = self.conditions_for_rule(&raw.id).await?;
            let actions = self.actions_for_rule(&raw.id).await?;
            rules.push(raw.into_rule(conditions, actions));
        }
        Ok(rules)
    }

    async fn list_for_overlay_inner(&self, user_id: &str) -> Result<Vec<OverlayRule>> {
        let rules = self.list_inner(user_id).await?;
        Ok(rules
            .into_iter()
            .map(|r| OverlayRule {
                logic: r.logic,
                conditions: r.conditions.iter().map(cond_to_data).collect(),
                actions: r.actions,
            })
            .collect())
    }

    async fn conditions_for_rule(&self, rule_id: &str) -> Result<Vec<RuleCondition>> {
        let rows = sqlx::query_as::<_, RawCondition>(
            "SELECT match_field, operator, pattern FROM rule_conditions WHERE rule_id = ?",
        )
        .bind(rule_id)
        .fetch_all(&self.pool)
        .await?;
        Ok(rows.into_iter().map(Into::into).collect())
    }

    async fn actions_for_rule(&self, rule_id: &str) -> Result<Vec<RuleAction>> {
        let rows = sqlx::query_as::<_, RawAction>(
            "SELECT action_type, name_op, value, category_id, transfer_account_id FROM rule_actions WHERE rule_id = ? ORDER BY rowid",
        )
        .bind(rule_id)
        .fetch_all(&self.pool)
        .await?;
        Ok(rows.into_iter().map(Into::into).collect())
    }
}

#[async_trait]
impl crate::repo::traits::RuleRepo for SqliteRuleRepo {
    async fn create(&self, user_id: &str, name: &str, priority: i64, logic: RuleLogic, conditions: &[RuleCondition], actions: &[RuleAction]) -> Result<Rule> {
        self.create_inner(user_id, name, priority, logic, conditions, actions).await
    }
    async fn update(&self, user_id: &str, id: &str, name: &str, priority: i64, logic: RuleLogic, conditions: &[RuleCondition], actions: &[RuleAction]) -> Result<bool> {
        self.update_inner(user_id, id, name, priority, logic, conditions, actions).await
    }
    async fn delete(&self, user_id: &str, id: &str) -> Result<bool> {
        self.delete_inner(user_id, id).await
    }
    async fn get_by_id(&self, user_id: &str, id: &str) -> Result<Option<Rule>> {
        self.get_by_id_inner(user_id, id).await
    }
    async fn list(&self, user_id: &str) -> Result<Vec<Rule>> {
        self.list_inner(user_id).await
    }
    async fn list_for_overlay(&self, user_id: &str) -> Result<Vec<OverlayRule>> {
        self.list_for_overlay_inner(user_id).await
    }
}

fn cond_to_data(c: &RuleCondition) -> ConditionData {
    ConditionData { match_field: c.match_field, operator: c.operator, pattern: c.pattern.clone() }
}

async fn insert_conditions(tx: &mut sqlx::Transaction<'_, sqlx::Sqlite>, rule_id: &str, conditions: &[RuleCondition]) -> Result<()> {
    for cond in conditions {
        sqlx::query("INSERT INTO rule_conditions (id, rule_id, match_field, operator, pattern) VALUES (?, ?, ?, ?, ?)")
            .bind(Uuid::new_v4().to_string())
            .bind(rule_id)
            .bind(cond.match_field.to_string())
            .bind(cond.operator.to_string())
            .bind(&cond.pattern)
            .execute(&mut **tx)
            .await?;
    }
    Ok(())
}

async fn insert_actions(tx: &mut sqlx::Transaction<'_, sqlx::Sqlite>, rule_id: &str, actions: &[RuleAction]) -> Result<()> {
    for act in actions {
        let (action_type, name_op, value, cat_id, transfer_id) = classify_action(act);
        sqlx::query(
            "INSERT INTO rule_actions (id, rule_id, action_type, name_op, value, category_id, transfer_account_id) VALUES (?, ?, ?, ?, ?, ?, ?)",
        )
        .bind(Uuid::new_v4().to_string())
        .bind(rule_id)
        .bind(action_type.to_string())
        .bind(name_op)
        .bind(value)
        .bind(cat_id)
        .bind(transfer_id)
        .execute(&mut **tx)
        .await?;
    }
    Ok(())
}

fn classify_action(act: &RuleAction) -> (ActionType, String, String, String, String) {
    if !act.set_transfer_account_id.is_empty() {
        (ActionType::SetTransferAccount, String::new(), String::new(), String::new(), act.set_transfer_account_id.clone())
    } else if !act.set_category_id.is_empty() {
        (ActionType::SetCategory, String::new(), String::new(), act.set_category_id.clone(), String::new())
    } else {
        let op = act.set_name_op.unwrap_or(ActionOp::Rename);
        (ActionType::SetName, op.to_string(), act.set_name.clone(), String::new(), String::new())
    }
}

#[derive(sqlx::FromRow)]
struct RawRule {
    id: String,
    name: String,
    priority: i64,
    logic: String,
    created_at: String,
}

impl RawRule {
    fn into_rule(self, conditions: Vec<RuleCondition>, actions: Vec<RuleAction>) -> Rule {
        Rule {
            id: self.id,
            name: self.name,
            priority: self.priority,
            logic: parse_logic(&self.logic),
            conditions,
            actions,
            created_at: ts_rfc3339(&self.created_at),
        }
    }
}

fn parse_logic(s: &str) -> RuleLogic {
    match s {
        "AND" => RuleLogic::And,
        _ => RuleLogic::Or,
    }
}

#[derive(sqlx::FromRow)]
struct RawCondition {
    match_field: String,
    operator: String,
    pattern: String,
}

impl From<RawCondition> for RuleCondition {
    fn from(r: RawCondition) -> Self {
        Self {
            match_field: r.match_field.parse().unwrap_or(MatchField::Name),
            operator: r.operator.parse().unwrap_or(MatchOperator::Contains),
            pattern: r.pattern,
        }
    }
}

#[derive(sqlx::FromRow)]
struct RawAction {
    action_type: String,
    name_op: String,
    value: String,
    category_id: String,
    transfer_account_id: String,
}

impl From<RawAction> for RuleAction {
    fn from(r: RawAction) -> Self {
        match r.action_type.parse::<ActionType>().unwrap_or(ActionType::SetName) {
            ActionType::SetTransferAccount => RuleAction { set_transfer_account_id: r.transfer_account_id, ..Default::default() },
            ActionType::SetCategory => RuleAction { set_category_id: r.category_id, ..Default::default() },
            ActionType::SetName => RuleAction {
                set_name: r.value,
                set_name_op: r.name_op.parse().ok(),
                ..Default::default()
            },
        }
    }
}
